File: Java/BinaryTrees/Level Order Traversal.java
package BinaryTrees;

import java.util.ArrayList;
import java.util.List;

class QueueNode<T> {
    public T getData() {
        return data;
    }

    public QueueNode<T> getNext() {
        return next;
    }

    private T data;
    private QueueNode<T> next;

    public QueueNode(T data) {
        this.data = data;
        this.next = null;
    }

    public void setNext(QueueNode<T> next) {
        this.next = next;
    }
}

class Queue<T> {
    private QueueNode<T> head;
    private QueueNode<T> tail;
    private Integer length;

    public Queue() {
        this.head = this.tail = null;
        this.length = 0;
    }

    public boolean isEmpty() {
        return length == 0;
    }

    public void push(T x) {
        QueueNode<T> node = new QueueNode<>(x);
        if (isEmpty()) {
            head = tail = node;
        }
        else {
            tail.setNext(node);
            tail = node;
        }
        length += 1;
    }

    public T pop() {
        if (isEmpty()) {
            return null;
        }
        T item = head.getData();
        head = head.getNext();
        length -= 1;
        return item;
    }
}


class Node<T> {
    private T data;
    private Node<T> left;
    private Node<T> right;

    public void setLeft(Node<T> left) {
        this.left = left;
    }

    public void setRight(Node<T> right) {
        this.right = right;
    }

    public T getData() {
        return data;
    }

    public Node<T> getLeft() {
        return left;
    }

    public Node<T> getRight() {
        return right;
    }

    public Node(T data) {
        this.data = data;
        this.left = this.right = null;
    }
}


class Solution {
    public static void main(String[] args) {
        // Example 1
        Node<Integer> node1 = new Node<>(1);
        Node<Integer> node2 = new Node<>(2);
        Node<Integer> node3 = new Node<>(3);
        node1.setLeft(node2);
        node1.setRight(node3);
        System.out.println(getLevelOrderTraversal(node1));

        // Example 2
        Node<Integer> n10 = new Node<>(10);
        Node<Integer> n20 = new Node<>(20);
        Node<Integer> n30 = new Node<>(30);
        Node<Integer> n40 = new Node<>(40);
        Node<Integer> n50 = new Node<>(50);
        n10.setLeft(n20);
        n10.setRight(n30);
        n20.setLeft(n40);
        n20.setRight(n50);
        System.out.println(getLevelOrderTraversal(n10));

        // Example 3
        Node<Integer> n1 = new Node<>(1);
        Node<Integer> n2 = new Node<>(2);
        Node<Integer> n3 = new Node<>(3);
        Node<Integer> n4 = new Node<>(4);
        Node<Integer> n5 = new Node<>(5);
        Node<Integer> n6 = new Node<>(6);
        n1.setLeft(n3);
        n1.setRight(n2);
        n2.setRight(n4);
        n4.setLeft(n6);
        n4.setRight(n5);
        System.out.println(getLevelOrderTraversal(n1));
    }

    public static <T> List<T> getLevelOrderTraversal(Node<T> root) {
        Queue<Node<T>> queue = new Queue<>();
        queue.push(root);
        List<T> result = new ArrayList<>();
        while (!queue.isEmpty()) {
            Node<T> node = queue.pop();
            result.add(node.getData());
            if (node.getLeft() != null) {
                queue.push(node.getLeft());
            }
            if (node.getRight() != null) {
                queue.push(node.getRight());
            }
        }
        return result;
    }
}

File: Python/Practice Set 1/Binary Trees/level order traversal.py
class QueueNode:
    def __init__(self, data):
        self.data = data
        self.next = None


class Queue:
    def __init__(self):
        self.head = self.tail = None
        self.length = 0

    def is_empty(self):
        return self.length == 0

    def push(self, x):
        node = QueueNode(x)
        if self.is_empty():
            self.head = self.tail = node
        else:
            self.tail.next = node
            self.tail = node
        self.length += 1

    def pop(self):
        if self.is_empty():
            return
        item = self.head.data
        node = self.head
        self.head = self.head.next
        del node
        self.length -= 1
        return item


class Node:
    def __init__(self, data):
        self.data = data
        self.left = self.right = None
        

class Solution:
    @staticmethod
    def get_level_order_traversal(root: Node):
        level_order_traversal = {}
        if root is None:
            return level_order_traversal
        queue = Queue()
        queue.push((root, 0))
        while not queue.is_empty():
            node, level = queue.pop()
            if level not in level_order_traversal:
                level_order_traversal[level] = [node.data,]
            else:
                level_order_traversal[level].append(node.data)
            if node.left is not None:
                queue.push((node.left, level + 1))
            if node.right is not None:
                queue.push((node.right, level + 1))
        return level_order_traversal
    

# Example 1
one, two, three = Node(1), Node(2), Node(3)
one.left = three
one.right = two
print(Solution.get_level_order_traversal(one))

# Example 2
ten, twty, thrty, fourty, sixty, svnty = Node(10), Node(20), Node(30), Node(40), Node(60), Node(70)
ten.left = twty
twty.left = fourty
twty.right = sixty
ten.right = thrty
thrty.right = svnty
print(Solution.get_level_order_traversal(ten))